File: examples/item-list/src/main/java/com/gurunars/item_list/example/AnimalItem.java
package com.gurunars.item_list.example;

import com.gurunars.item_list.Item;

class AnimalItem extends Item {

    private int version;

    @Override
    public boolean payloadsEqual(Item other) {
        return other instanceof AnimalItem && version == ((AnimalItem) other).version;
    }

    enum Type {
        MONKEY, TIGER, WOLF, LION
    }

    void update() {
        this.version++;
    }

    AnimalItem(long id, int version, Type type) {
        super(id, type);
        this.version = version;
    }

    AnimalItem(long id, int version) {
        this(id, version, Type.MONKEY);
    }

    AnimalItem(long id, Type type) {
        this(id, 0, type);
    }

    @Override
    public String toString() {
        return "#" + getId() + "{" + getType() + " @ " + version + "}";
    }

}

File: examples/item-list/src/main/java/com/gurunars/item_list/example/AnimalItem.kt
package com.gurunars.item_list.example

import com.gurunars.item_list.Item

internal class AnimalItem(id: Long, private var version: Int, type: AnimalItem.Type) : Item(id, type) {

    override fun payloadsEqual(other: Item): Boolean {
        return other is AnimalItem && version == other.version
    }

    internal enum class Type {
        MONKEY, TIGER, WOLF, LION
    }

    fun update() {
        this.version++
    }

    override fun toString(): String {
        return "#$id{$type @ $version}"
    }

}
